File: src/exemplos/05_EstruturasDeDados/03_Ponteiros/03-ponteiro.py
#  -*- coding: utf-8 -*-
#    @package: 03-ponteiro.py
#     @author: Guilherme N. Ramos (gnramos@unb.br)
# @disciplina: Algoritmos e Programação de Computadores
#
# Comportamento com referências (ponteiros).

# Python sempre realiza passagem por referência, não há exceção. Qualquer
# atribuição significa copiar o valor da referência.

# Exemplo:

# Inicializa vetor de 5 elementos.
vetor = [0.0 for _ in range(5)]

# Mostra cada elemento de vetor
for elemento in vetor:
    print(elemento)
# 0.0
# 0.0
# 0.0
# 0.0
# 0.0


# Copiamos a referência de vetor para uma outra variável
n = vetor

# Modificamos alguns elemento
n[0] = 1.0
n[3] = 3.14

# Verificando a situação do vetor
for elemento in vetor:
    print(elemento)
#  1.0
#  0.0
#  0.0
#  0.0
#  0.0

# Qual seria uma forma correta de copiar todos os valores para n e realizar as
# alterações sem modificar o vetor?

File: src/exemplos/05_EstruturasDeDados/03_Ponteiros/03-ponteiro.c
/**      @file: 03-ponteiro.c
 *     @author: Guilherme N. Ramos (gnramos@unb.br)
 * @disciplina: Algoritmos e Programação de Computadores
 *
 * Exemplo de uso de ponteiro para acessar escopos. */

#include <stdio.h>
#include "apc_ponteiro.h"

/* O arquivo apc_ponteiros define a função troca_i que troca os valores de x e y
 * no escopo dado. */

int main() {
    int x = 1;
    int y = 2;

    printf("x = %d\n", x);
    printf("y = %d\n", y);
    printf("Trocando...\n");
    troca_i(&x, &y);
    printf("Trocou!\n");
    printf("x = %d\n", x);
    printf("y = %d\n", y);

    return 0;
}